### ISSUE TITLE
Running semgrep on Apple M1 ARM system

### ISSUE THREAD

**Issue by mschwager** [semgrep/semgrep#2432]
The new [Apple M1](https://en.wikipedia.org/wiki/Apple_M1) chips are ARM-based. Currently, our OCaml binaries are compiled for x86-64. It seems the new M1 chips have [x86 emulation under something called Rosetta 2](https://wccftech.com/m1-macbook-air-x86-emulation-faster-all-mac-models/), so things may Just Work. Regardless, we should test things out to ensure they do.

**Comment by underyx** [comment]
I believe @fallion (@capchase) is using an M1 chip and he might have already been running Semgrep on his machine.

**Comment by aryx** [comment]
OCaml has also been recently ported to this new archi/os, so we should be able to produce native binaries for it.

**Comment by aryx** [comment]
related also: https://github.com/returntocorp/semgrep/issues/2252

**Comment by DrewDennison** [comment]
r2c doesn't have an new M1 macs yet 😢 but I believe we could test this with arm docker images but would love a PR or help from anyone in the community with ARM compilation experience. 🙏

**Comment by brendongo** [comment]
Haven't tested on an M1 mac yet but 0.43.0 on homebrew included building on ARM https://github.com/Homebrew/homebrew-core/pull/73280

**Comment by aryx** [comment]
related slack issue? https://r2c-community.slack.com/archives/CK86BJ5DW/p1617739952157800

**Comment by brendongo** [comment]
Semgrep on Rosetta x86-64 emulation seems to be working fine but the ARM build from homebrew seems stuck when running on juice-shop and p/security-audit

**Comment by thypon** [comment]
Same error here

**Comment by tino** [comment]
How can I build semgrep-core on an M1? I have no experience with OCaml.

We've set up semgrep with pre-commit, but I run into:
```
OSError: [Errno 8] Exec format error: '/Users/tino/.cache/pre-commit/repo_6jxvknx/py_env-python3.9/lib/python3.9/site-packages/semgrep/bin/semgrep-core'
```

If I know how to build semgrep-core I can swap that out and see if it works.

**Comment by aryx** [comment]
Hopefully @tino we can build an homebrew m1-compatible package very soon.

Regarding compiling OCaml projects on M1, according to https://discuss.ocaml.org/t/ocaml-for-arm-macos/6019/23 the OCaml package manager OPAM now handles correctly M1.
So brew install opam should work
From there forllow the instructions in CONTRIBUTING.md to compile semgrep-core.

**Comment by tino** [comment]
I managed to do that, and it works. However I symlinked it into the pre-commit install but now it always says `running 0 rules...`. 

So I guess I'll wait for the build :)

**Comment by brendongo** [comment]
Updates and clarifications:

tl;dr: On M1 Macs install semgrep in Rosetta using pip not brew

On Rosetta I am able to run semgrep when it was installed with pip:
```
python3 -m pip install semgrep
python3 -m semgrep --config p/security-audit juice-shop
```
but on Rosetta, when using the homebrew installed semgrep I get stuck on a rule with semgrep-core using up 100% CPU

Without Rosetta, the pip installed version doesn't. The pip installed one not working is expected since we do not publish an ARM binary in pypi right now. The brew install also stalls outside Rosetta.

**Comment by tino** [comment]
So building from source on M1 (without Rosetta) works! I'm not sure how to invoke it (as there is only semgrep-core), but I fixed my pre-commit by symlinking semgrep-core into the pre-commit virtualenv. No CPU spikes here.

So how can we build an ARM wheel that includes such a build?

**Comment by thypon** [comment]
So `brew install --build-from-source semgrep` will already compile and install semgrep for the hosting architecture (Apple M1 on M1).

Can you confirm that is working in that case?

**Comment by brendongo** [comment]
@thypon when running`brew install --build-from-source semgrep` on Apply M1 on M1 i still have `semgrep --config p//security-audit juice-shop` not making progress

**Comment by thypon** [comment]
Me too. It seems to me that the bug is in the python wrapper and not in the core thought. On some rules the python wrapper memory bombs my machine until the out-of-memory killer kills it. No matter if it is running native M1 or Rosetta.

**Comment by brendongo** [comment]
Hmm I'm seeing semgrep-core jobs not exiting so the python wrapper is left waiting for the subprocess call to semgrep-core to return. But yeah looks like just specific rules are causing this. When I get some time I'll try and track down which specific rules those are to make this easier to debug :D

**Comment by hex0punk** [comment]
This rules is hanging on M1. I haven't done much to debut it yet, but I know the issue does not occur in Intel Macs: https://github.com/trailofbits/semgrep-rules/blob/main/go/anonymous-race-condition.yml

**Comment by IagoAbal** [comment]
I can build from sources ("manually" rather than using `brew`) and run the tests in semgrep-rules without problems on M1. I can also run our perf benchmark that includes real-world code without problems. Can you point to a concrete rule and file that causes these problems?

**Comment by brendongo** [comment]
Back on this. Can confirm that building from source runs without problems on M1 but running `brew install -s semgrep` runs into problems

**Comment by brendongo** [comment]
Well I have a weird minimal reproduction of the issue:

1. Install using brew build from source: `brew install -s semgrep`
2. Create a test directory and run the following commands:
```
echo 'print(1)' > 1.py
echo 'print(1)' > 2.py
echo 'print(1)' > 3.py
echo 'print(1)' > 4.py
echo 'print(1)' > 5.py
echo 'print(1)' > 6.py
echo 'print(1)' > 7.py
echo 'print(1)' > 8.py
```
3. Run `semgrep -e 'print($X)' --lang python .` This should run to completion without issue
4. Run `cp 1.py 9.py`
5. Run `semgrep -e 'print($X)' --lang python .` Semgrep will no longer finish


This seems to suggest there is some race condition or other issue on the multiprocessing part of semgrep-core

With 9 files, running `semgrep -e 'print($X)' -j 4 --lang python .` exits without issue but for some reason `semgrep -e 'print($X)' -j 5 --lang python .`  no longer makes progress. (Am running on an 8 core M1 MacBook Air)

Still need to look into:
- Does this work around work for juice-shop with security-audit rules
- why this behavior exists when `brew install -s semgrep` but not when building from source outside brew
- whether this behavior exists when `brew install semgrep` without building from source

**Comment by brendongo** [comment]
Running the built semgrep-core binaries directly:

Infinite loop is printing out: 
```
MAC OS X: Failed pinning to cpu 0, trying 0/2
MAC OS X: Failed pinning to cpu 0, trying 0/2
MAC OS X: Failed pinning to cpu 0, trying 0/2
```

Which seems linked to parmap https://github.com/rdicosmo/parmap/blob/f140dbc8b05aa3d341c70436a1920a06df9a0ed4/src/setcore_stubs.c#L54 which supports @IagoAbal hypothesis on an unpinned ocaml dependency being the potential problem? But could also point to diff setting of HAVE_MACH_THREAD_POLICY_H in brew builld env and outside brew build env.

I wonder if we can bypass the core pinning by using this flag https://github.com/rdicosmo/parmap/pull/69/files#diff-a54c8964b69da8bc2554c58f9c0bcea90e05acc269ef30eab84a6d6597648dbcR40

Related ticket on infer: https://github.com/facebook/infer/issues/1410

The infinite loop occurs when -j is set to 5,6,7, or 8

**Comment by brendongo** [comment]
Acceptable workarounds so far:

1. Use Rosetta and use pypi semgrep
2. Use brew installed semgrep but set `--jobs` to 1 when running semgrep
3. Follow steps in https://github.com/returntocorp/semgrep/blob/develop/doc/SEMGREP_CORE_CONTRIBUTING.md#building-semgrep-core to build semgrep-core from source on an M1 Mac

**Comment by thypon** [comment]
Is there a roadmap to provide a bug-free experience for the general public?

**Comment by IagoAbal** [comment]
@thypon yes, I will be looking into this this week.

**Comment by IagoAbal** [comment]
Simply `Parmap.disable_core_pinning();` fixes the problem in 0.56.0, I will check this against develop and I'll open a PR. The fix should go into 0.58.0 next week (it's already too late to include it in 0.57.0).

**Comment by brendongo** [comment]
For people running into issues state of the world as of 0.61.0 is on M1 Mac the following should work:
- brew install 
- pypi install with rosetta

**Comment by pranto157** [comment]
Hi I am having this issue with my new M1. Any idea how can I fix it?

**Comment by raianul** [comment]
@brendongo 

I am running with - MacOs Monterey 
M1 Chip 

After try all the option still getting error - 

```    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/opt/homebrew/Cellar/python@3.10/3.10.2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/subprocess.py", line 1842, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
OSError: [Errno 8] Exec format error: '/Users/sayed.kabir/.cache/pre-commit/repovfpqfv44/py_env-python3.10/lib/python3.10/site-packages/semgrep/bin/semgrep-core'```

**Comment by raianul** [comment]
```➜  shorepay-service git:(update-precomit-config) ✗ semgrep --config=semgrep_rules.yaml
Scanning 126 files.
    0%|                                                                                                                                       |0/126 tasksTraceback (most recent call last):
  File "/Users/sayed.kabir/.pyenv/versions/3.10.3/bin/semgrep", line 8, in <module>
    sys.exit(main())
  File "/Users/sayed.kabir/.pyenv/versions/3.10.3/lib/python3.10/site-packages/semgrep/__main__.py", line 20, in main
    cli()
  File "/Users/sayed.kabir/.pyenv/versions/3.10.3/lib/python3.10/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/Users/sayed.kabir/.pyenv/versions/3.10.3/lib/python3.10/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/Users/sayed.kabir/.pyenv/versions/3.10.3/lib/python3.10/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/sayed.kabir/.pyenv/versions/3.10.3/lib/python3.10/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/sayed.kabir/.pyenv/versions/3.10.3/lib/python3.10/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/Users/sayed.kabir/.pyenv/versions/3.10.3/lib/python3.10/site-packages/semgrep/commands/scan.py", line 868, in scan
    ) = semgrep.semgrep_main.main(
  File "/Users/sayed.kabir/.pyenv/versions/3.10.3/lib/python3.10/site-packages/semgrep/semgrep_main.py", line 340, in main
    rule_matches_by_rule, semgrep_errors, all_targets, profiling_data = run_rules(
  File "/Users/sayed.kabir/.pyenv/versions/3.10.3/lib/python3.10/site-packages/semgrep/semgrep_main.py", line 146, in run_rules
    ) = core_runner.invoke_semgrep(
  File "/Users/sayed.kabir/.pyenv/versions/3.10.3/lib/python3.10/site-packages/semgrep/core_runner.py", line 680, in invoke_semgrep
    ) = self._run_rules_direct_to_semgrep_core(
  File "/Users/sayed.kabir/.pyenv/versions/3.10.3/lib/python3.10/site-packages/semgrep/core_runner.py", line 626, in _run_rules_direct_to_semgrep_core
    returncode = runner.execute()
  File "/Users/sayed.kabir/.pyenv/versions/3.10.3/lib/python3.10/site-packages/semgrep/core_runner.py", line 249, in execute
    rc = asyncio.run(self._stream_subprocess())
  File "/Users/sayed.kabir/.pyenv/versions/3.10.3/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/Users/sayed.kabir/.pyenv/versions/3.10.3/lib/python3.10/asyncio/base_events.py", line 646, in run_until_complete
    return future.result()
  File "/Users/sayed.kabir/.pyenv/versions/3.10.3/lib/python3.10/site-packages/semgrep/core_runner.py", line 208, in _stream_subprocess
    process = await asyncio.create_subprocess_exec(
  File "/Users/sayed.kabir/.pyenv/versions/3.10.3/lib/python3.10/asyncio/subprocess.py", line 218, in create_subprocess_exec
    transport, protocol = await loop.subprocess_exec(
  File "/Users/sayed.kabir/.pyenv/versions/3.10.3/lib/python3.10/asyncio/base_events.py", line 1667, in subprocess_exec
    transport = await self._make_subprocess_transport(
  File "/Users/sayed.kabir/.pyenv/versions/3.10.3/lib/python3.10/asyncio/unix_events.py", line 207, in _make_subprocess_transport
    transp = _UnixSubprocessTransport(self, protocol, args, shell,
  File "/Users/sayed.kabir/.pyenv/versions/3.10.3/lib/python3.10/asyncio/base_subprocess.py", line 36, in __init__
    self._start(args=args, shell=shell, stdin=stdin, stdout=stdout,
  File "/Users/sayed.kabir/.pyenv/versions/3.10.3/lib/python3.10/asyncio/unix_events.py", line 799, in _start
    self._proc = subprocess.Popen(
  File "/Users/sayed.kabir/.pyenv/versions/3.10.3/lib/python3.10/subprocess.py", line 966, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/Users/sayed.kabir/.pyenv/versions/3.10.3/lib/python3.10/subprocess.py", line 1842, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
OSError: [Errno 8] Exec format error: '/Users/sayed.kabir/.pyenv/versions/3.10.3/lib/python3.10/site-packages/semgrep/bin/semgrep-core'
    0%|

**Comment by brendongo** [comment]
@pranto157 looks like on the first one you are running in a pre-commit context so not sure what is going on there but homebrew compiles on ARM so it is possible the pre-commit context is trying to emulate x86

The second looks like running in pyenv was this virtualenv running in rosetta? I would expect a pip install semgrep on M1 to only work if x86 emulation is on.

**Comment by rohitcoder** [comment]
Any plans for this? The release was expected in mid of May 2022, I've had to switch from my mac to an Ubuntu server every time to work on semgrep, the worst part is VirtualBox also does not work with M1. 

If someone is having any workarounds please let me know

**Edit:**
As a workaround, I managed to run this using docker on my Mac M1 

```
sudo docker run --rm -v $(pwd):/src returntocorp/semgrep semgrep --config=semgrep-rules/python projPy --verbose
 ```

**Comment by aryx** [comment]
maybe @spencerdrak has some idea on how to fix the issue?

**Comment by brendongo** [comment]
As of 0.96.0 an M1 compatible version of semgrep is released on pypi. Considering this issue closed now that brew and pypi work on M1.

**Comment by brendongo** [comment]
@rohitcoder fyi ^

**Comment by aryx** [comment]
Cc @ajinabraham

**Comment by ajinabraham** [comment]
Awesome
